### DISCUSSION TITLE
v0.11.4 Testing checklist

### DISCUSSION THREAD

**Discussion by jorgeepc** [kubeshop/tracetest#2377] [Release testing]
# Testing Tracetest

This is the entire checklist on what we should do to assert that Tracetest is working fine on each version release. On each version release, we can copy the contents of this template checklist and open a Github Discussion to start each test.

On Tracetest, we work in two ways to test the entire system and guarantee that everything is working fine. One is through automatic tests, where we run unit, integration and end-to-end tests on Tracetest CLI, Web UI and Server.

Another source of tests is the manual tests that we execute on each release, simulating users, and checking if everything is ok on these tests.

## Manual Tests

On our manual tests, we aim to do some [sanity checks](https://en.wikipedia.org/wiki/Sanity_check) to see if the main features are working as expected. Usually, we run it on each Tracetest release.

### Tests to validate on Release Candidate

- Test server installation via CLI
  - [x] Docker Compose and no demo API
  - [x] Docker Compose and demo API
  - [x] Kubernetes and no demo API
  - [x] Kubernetes and demo API
  - [ ] ~~Verify that Tracetest is not sending analytics if you have the TRACETEST_DEV env~~

- Test DataStore provisioning
  - [x] Open WebUI and go to `/settings` page. The provisioned Data Store should be selected
  - [x] Create a test on WebUI that calls a demo API (like [Pokeshop](https://docs.tracetest.io/live-examples/pokeshop/overview) or [Open Telemetry Store](https://docs.tracetest.io/live-examples/opentelemetry-store/overview)). This test should fetch traces correctly and run without errors

- Test Tracetest examples (we haven't changed code related to this, we can skip testing it)
  - [ ] ~~[Amazon X-Ray example](https://github.com/kubeshop/tracetest/tree/main/examples/tracetest-amazon-x-ray)~~
  - [ ] ~~[Datadog example](https://github.com/kubeshop/tracetest/tree/main/examples/tracetest-datadog)~~
  - [ ] ~~[Elastic APM example](https://github.com/kubeshop/tracetest/tree/main/examples/tracetest-elasticapm)~~
  - [ ] ~~[Lightstep example](https://github.com/kubeshop/tracetest/tree/main/examples/tracetest-lightstep)~~
  - [ ] ~~[New Relic example](https://github.com/kubeshop/tracetest/tree/main/examples/tracetest-new-relic)~~
  - [ ] ~~[SignalFX example](https://github.com/kubeshop/tracetest/tree/main/examples/tracetest-signalfx)~~

- Test specific features added/changed on this release:
  - [x] Create a test and add Test specs using snippets
  - [x] Validate the new server analytic `test_run_finished`
  
### Tests to validate on Final Release

- Pipeline
  - [x] Check if our release pipeline on [Release Tracetest](https://github.com/kubeshop/tracetest/actions/workflows/release-version.yml) workflow on Github Actions worked correctly

- Test CLI updates
  - [x] MacOS via homebrew
  - [x] MacOS via curl script
  - [x] Windows via chocolatey
  - [x] Windows via manual download

- Docs
  - [x] Double check [Detailed installation](https://docs.tracetest.io/getting-started/detailed-installation) docs and see if everything is documented correctly

- Test specific features added/changed on this release:
  - [x] Create a test and add Test specs using snippets
  - [x] Validate the new server analytic `test_run_finished`
  
## Automatic Tests

Today Tracetest has 3 main components: a WebUI, a CLI and a Server.

### Web UI

- **Unit tests**: Run by executing `npm test` on `./web` folder
- **End-to-end tests**: Run using [cypress](https://www.cypress.io/) against a temporary Tracetest created on Kubernetes. 

### CLI

- **Unit tests**: Run by executing `make test` on `./cli` folder

### Server

- **Unit tests**: Run by executing `make test` on `./server` folder
- **Integration tests**: Run along with some unit tests running `make test` on `./server` folder, but also done in a matrix test on Github actions, by executing the `test-examples` action.
- **End-to-end tests**: Run using Tracetest to test itself. We deploy two instances of Tracetest and use one to test API calls to another on Github actions, by executing the `trace-testing` action.

**Comment by xoscar** [comment]
All done!
